bluesky/plans.py

```diff
@@ -90,36 +96,67 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+            raise ValueError("The list of arguments must contain a list of"
+                             "points for each defined motor")
+
+    # check that all lists are the same length
+    lengths = {}
+    motors = []
+    pos_lists = []
+    for motor, pos_list in partition(2,args):
+        lengths[motor.name] = len(pos_list)
+        motors.append(motor)
+        pos_lists.append(pos_list)
+    lengths = {motor.name: len(pos_list) for motor, pos_list
+               in partition(2,args)}
+    length_check = all(elem == list(lengths.values()[0]) for elem in
+                       list(lengths.values()))
+
+    if not length_check:
+        raise ValueError("The lengths of all lists in *args must be the same"
+                         f"However the lengths in args are : {lengths}")
+
+    md_args = list(chain(*((repr(motor), pos_list)
+                           for motor, pos_list in partition(2, args))))
+    motor_names = list(lengths.keys())
+
     _md = {'detectors': [det.name for det in detectors],
-           'motors': [motor.name],
-           'num_points': len(steps),
-           'num_intervals': len(steps) - 1,
+           'motors': motor_names,
+           'num_points': len(pos_lists[0]),
+           'num_intervals': len(pos_lists[0]) - 1,
            'plan_args': {'detectors': list(map(repr, detectors)),
-                         'motor': repr(motor), 'steps': steps,
+                         'args': md_args,
                          'per_step': repr(per_step)},
            'plan_name': 'list_scan',
            'plan_pattern': 'array',
            'plan_pattern_module': 'numpy',
-           'plan_pattern_args': dict(object=steps),
+           'plan_pattern_args': lengths,
            'hints': {},
            }
     _md.update(md or {})
-    try:
-        dimensions = [(motor.hints['fields'], 'primary')]
-    except (AttributeError, KeyError):
-        pass
-    else:
-        _md['hints'].setdefault('dimensions', dimensions)
-    if per_step is None:
-        per_step = bps.one_1d_step
 
-    @bpp.stage_decorator(list(detectors) + [motor])
-    @bpp.run_decorator(md=_md)
-    def inner_list_scan():
-        for step in steps:
-            yield from per_step(detectors, motor, step)
+    x_fields = []
+    for motor in motors:
+        x_fields.extend(getattr(motor, 'hints', {}).get('fields', []))
+
+    default_dimensions = [(x_fields, 'primary')]
+
+    default_hints = {}
+    if len(x_fields) > 0:
+        default_hints.update(dimensions=default_dimensions)
+
+    # now add default_hints and override any hints from the original md (if
+    # exists)
+    _md['hints'] = default_hints
+    _md['hints'].update(md.get('hints', {}) or {})
+
+    cycler = []
+    for i in range(0, len(pos_lists[0])):
+       cycler.append({motor: pos_list[i]
+                      for motor, pos in zip(motors, pos_lists)})
 
-    return (yield from inner_list_scan())
+    return (yield from scan_nd(detectors, cycler, per_step=per_step, md=_md))
```

Comment:
This should be a `cycler` object, not a list of `cycler` objects. I believe that above , where you define the cycler, you want something like:

```python
cycler = sum(cycler(motor, pos_list) for motor, pos_list in partition(2,args))
```

Reply:
thanks I have made this change